In [1]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler, PolynomialFeatures
from sklearn.model_selection import train_test_split, cross_val_score, KFold
from sklearn.linear_model import LinearRegression, Ridge, Lasso, ElasticNet, BayesianRidge
from sklearn.svm import SVR
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.neighbors import KNeighborsRegressor
from sklearn.neural_network import MLPRegressor
from sklearn.pipeline import make_pipeline
from sklearn.decomposition import PCA
from sklearn.cross_decomposition import PLSRegression
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
import joblib  # Correct import for joblib
from xgboost import XGBRegressor
import matplotlib.pyplot as plt

# Load Dataset

In [2]:
# Load the final feature engineered data
df_final = pd.read_csv('final_feature_engineered_data.csv')

In [3]:
df_final

Day         Month          Year  Conc. Of  pH        EC    Bicarb  \
0   0.213766  1.110223e-16 -8.881784e-16      0.181918  0.529605  0.937836   
1   0.198689  1.110223e-16 -8.881784e-16      0.189731  0.521134  0.922757   
2   0.194115  1.110223e-16 -8.881784e-16      0.203582  0.512210  0.953746   
3   0.204799  1.110223e-16 -8.881784e-16      0.191872  0.549472  0.960658   
4   1.389907  1.110223e-16 -8.881784e-16     -0.409650 -0.473947 -0.564241   
5   1.389169  1.110223e-16 -8.881784e-16     -0.326212 -0.507226 -0.535738   
6   1.411476  1.110223e-16 -8.881784e-16     -0.329115 -0.501720 -0.586776   
7   1.402605  1.110223e-16 -8.881784e-16     -0.344410 -0.471387 -0.567157   
8   1.403965  1.110223e-16 -8.881784e-16     -0.961438 -0.985865 -0.578515   
9   1.443295  1.110223e-16 -8.881784e-16     -1.030846 -0.980619 -0.570541   
10  1.348919  1.110223e-16 -8.881784e-16     -0.987501 -1.009261 -0.564257   
11  1.413506  1.110223e-16 -8.881784e-16     -0.980599 -0.977828 -0.536462   
12 -1.191048  1.110223e-16 -8.881784e-16      0.742132  0.722244 -0.551269   
13 -1.163172  1.110223e-16 -8.881784e-16      0.765892  0.766761 -0.572920   
14 -1.148418  1.110223e-16 -8.881784e-16      0.714796  0.699271 -0.543553   
15 -1.139837  1.110223e-16 -8.881784e-16      0.772794  0.745214 -0.602549   
16  0.216617  1.110223e-16 -8.881784e-16      1.442548 -1.181108 -0.547141   
17  0.224547  1.110223e-16 -8.881784e-16      1.524893 -1.220027 -0.610562   
18  0.204835  1.110223e-16 -8.881784e-16      1.480022 -1.204236 -0.546412   
19  0.176050  1.110223e-16 -8.881784e-16      1.443563 -1.194565 -0.568110   
20  0.176482  1.110223e-16 -8.881784e-16      1.484093 -1.150029 -0.490397   
21 -1.648966  1.110223e-16 -8.881784e-16     -2.492794  1.620694  0.968468   
22 -1.648966  1.110223e-16 -8.881784e-16     -0.910556  0.061813  0.968468   
23 -1.648966  1.110223e-16 -8.881784e-16     -1.474684  1.511905  0.968468   
24 -1.156934  1.110223e-16 -8.881784e-16     -1.727366  2.376656  2.316229   
25 -1.156934  1.110223e-16 -8.881784e-16     -0.972938 -0.380401  0.968468   
26 -1.156934  1.110223e-16 -8.881784e-16      0.321770  0.315978  2.316229   
27 -1.156934  1.110223e-16 -8.881784e-16      0.806712  0.284894  0.968468   
28 -1.156934  1.110223e-16 -8.881784e-16      0.746339  0.703670 -0.558956   
29 -0.685845  1.110223e-16 -8.881784e-16     -0.106261  1.169755 -0.558956   
30 -0.685845  1.110223e-16 -8.881784e-16     -0.290777  1.090550 -0.558956   
31 -0.685845  1.110223e-16 -8.881784e-16      0.199831  1.375127  0.968468   
32  0.200140  1.110223e-16 -8.881784e-16      1.466277 -0.710772 -2.321421   
33  0.200140  1.110223e-16 -8.881784e-16     -0.414149 -0.983449  0.968468   
34  0.200140  1.110223e-16 -8.881784e-16     -0.910556 -0.864792 -0.558956   
35  0.200140  1.110223e-16 -8.881784e-16      1.466277 -1.187583 -0.558956   
36  0.200140  1.110223e-16 -8.881784e-16     -1.537740 -0.598246  0.968468   
37  0.200140  1.110223e-16 -8.881784e-16      0.260836 -0.344985  0.968468   
38  0.200140  1.110223e-16 -8.881784e-16      0.927251  2.314296 -2.321421   
39  0.200140  1.110223e-16 -8.881784e-16      0.199831  0.528402  0.968468   
40  1.408857  1.110223e-16 -8.881784e-16     -0.352427 -0.488156 -0.558956   
41  1.408857  1.110223e-16 -8.881784e-16     -0.972938 -0.983449 -0.558956   

       Chlor      Iron  Manganese    Sodium  Potassium  Magnesium  Log_Arsenic  
0  -0.897475 -0.332173  -0.199134 -0.990882  -0.718107   0.459290     2.712108  
1  -0.903200 -0.345305  -0.198272 -0.970444  -0.688951   0.477924     2.704008  
2  -0.860840 -0.374221  -0.185931 -0.971685  -0.696098   0.463224     2.671262  
3  -0.921168 -0.368752  -0.218919 -0.977194  -0.670931   0.455944     2.705551  
4  -1.396437 -0.353426  -0.788518 -1.319128   0.775543  -0.114794     2.811246  
5  -1.405846 -0.380960  -0.752810 -1.323950   0.745343  -0.129277     2.864956  
6  -1.402391 -0.333524  -0.767888 -1.328753   0.806713  -0.082678     2.827142  
7  -1.

# Defining The dependent and indepedent variable

In [4]:
# Separate the features and target
X = df_final.drop(columns=['Log_Arsenic'])
y = df_final['Log_Arsenic']

# Scaling for the final time

In [5]:
# Standardize the features
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

# Splitting the dataset

In [6]:
# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.2, random_state=42)

# Defining the model 

In [7]:
models = [
    ("Linear Regression", LinearRegression()),
    ("Polynomial Regression", PolynomialFeatures(degree=2)),  # This will need to be handled specially
    ("Ridge Regression", Ridge()),
    ("Lasso Regression", Lasso()),
    ("Elastic Net Regression", ElasticNet()),
    ("Support Vector Regression", SVR()),
    ("Decision Tree Regression", DecisionTreeRegressor()),
    ("Random Forest Regression", RandomForestRegressor(random_state=42)),
    ("Gradient Boosting Regression", GradientBoostingRegressor(random_state=42)),
    ("K-Nearest Neighbors Regression", KNeighborsRegressor()),
    ("Bayesian Regression", BayesianRidge()),
    ("Neural Network Regression", MLPRegressor(hidden_layer_sizes=(50,50,50), max_iter=1000, random_state=42)),
    ("Principal Component Regression (PCR)", make_pipeline(StandardScaler(), PCA(), LinearRegression())),
    ("Partial Least Squares Regression (PLS)", PLSRegression())
]
# Define evaluation metrics
metrics = {
    "MAE": mean_absolute_error,
    "MSE": mean_squared_error,
    "R2": r2_score
}

# Using the k-fold validation 

In [8]:
# Cross-validation
kf = KFold(n_splits=5, shuffle=True, random_state=42)

# Result 

In [9]:
# Dictionary to store the results
results = {}

# Iterate through the list of models
for name, model in models:
    print(f"Evaluating {name}...")
    
    if name == "Polynomial Regression":
        # Special handling for Polynomial Regression
        poly = PolynomialFeatures(degree=2)
        X_poly = poly.fit_transform(X_scaled)
        lr = LinearRegression()
        scores = cross_val_score(lr, X_poly, y, cv=kf, scoring='r2')
        lr.fit(X_poly, y)
        y_pred = lr.predict(poly.transform(X_test))
        model = lr  # Update model to be the trained Linear Regression
    elif name == "Principal Component Regression (PCR)":
        # Special handling for PCR
        pcr = make_pipeline(StandardScaler(), PCA(), LinearRegression())
        scores = cross_val_score(pcr, X_scaled, y, cv=kf, scoring='r2')
        pcr.fit(X_scaled, y)
        y_pred = pcr.predict(X_test)
        model = pcr  # Update model to be the trained pipeline
    else:
        # Standard handling for other models
        scores = cross_val_score(model, X_scaled, y, cv=kf, scoring='r2')
        model.fit(X_train, y_train)
        y_pred = model.predict(X_test)

    # Calculate metrics
    mae = mean_absolute_error(y_test, y_pred)
    mse = mean_squared_error(y_test, y_pred)
    r2 = r2_score(y_test, y_pred)
    
    # Store the results
    results[name] = {
        "CV_R2_Score": scores.mean(),
        "MAE": mae,
        "MSE": mse,
        "R2": r2
    }

    # Save the model
    joblib.dump(model, f"{name}.pkl")

# Create a DataFrame to display the results
results_df = pd.DataFrame(results).T
print(results_df)

# Save the results to a CSV file
results_df.to_csv('model_comparison_results.csv')

Evaluating Linear Regression...
Evaluating Polynomial Regression...
Evaluating Ridge Regression...
Evaluating Lasso Regression...
Evaluating Elastic Net Regression...
Evaluating Support Vector Regression...
Evaluating Decision Tree Regression...
Evaluating Random Forest Regression...
Evaluating Gradient Boosting Regression...
Evaluating K-Nearest Neighbors Regression...
Evaluating Bayesian Regression...
Evaluating Neural Network Regression...
Evaluating Principal Component Regression (PCR)...
Evaluating Partial Least Squares Regression (PLS)...
                                        CV_R2_Score           MAE  \
Linear Regression                          0.167647  4.561923e-01   
Polynomial Regression                      0.217806  9.128500e-16   
Ridge Regression                           0.317766  4.289743e-01   
Lasso Regression                          -0.311644  8.133298e-01   
Elastic Net Regression                    -0.288568  8.092209e-01   
Support Vector Regression          

# Here CV R2 Score means the k-fold cross Validation Score

In [23]:
results_df

CV_R2_Score           MAE  \
Linear Regression                          0.167647  4.561923e-01   
Polynomial Regression                      0.217806  9.128500e-16   
Ridge Regression                           0.317766  4.289743e-01   
Lasso Regression                          -0.311644  8.133298e-01   
Elastic Net Regression                    -0.288568  8.092209e-01   
Support Vector Regression                  0.846930  2.720220e-01   
Decision Tree Regression                   0.258534  4.560451e-01   
Random Forest Regression                   0.634661  4.198173e-01   
Gradient Boosting Regression               0.647519  4.061689e-01   
K-Nearest Neighbors Regression            -0.032279  4.831339e-01   
Bayesian Regression                        0.332541  4.352065e-01   
Neural Network Regression                  0.377827  4.236287e-01   
Principal Component Regression (PCR)       0.167647  2.062382e-01   
Partial Least Squares Regression (PLS)     0.348971  3.848405e-01   

                                                 MSE        R2  
Linear Regression                       4.337293e-01  0.518584  
Polynomial Regression                   1.922848e-30  1.000000  
Ridge Regression                        3.494912e-01  0.612083  
Lasso Regression                        9.699867e-01 -0.076633  
Elastic Net Regression                  9.534373e-01 -0.058264  
Support Vector Regression               1.931032e-01  0.785666  
Decision Tree Regression                7.352958e-01  0.183861  
Random Forest Regression                5.731295e-01  0.363857  
Gradient Boosting Regression            6.423270e-01  0.287051  
K-Nearest Neighbors Regression          6.104306e-01  0.322455  
Bayesian Regression                     3.334634e-01  0.629874  
Neural Network Regression               5.357772e-01  0.405316  
Principal Component Regression (PCR)    7.345748e-02  0.918466  
Partial Least Squares Regression (PLS)  3.102032e-01  0.655691

# OutPut Analysis 

## Observations:
 ### Polynomial Regression:

R2: 1.0, indicating a perfect fit on the test data.
MAE & MSE: Extremely low values (close to zero), indicating very accurate predictions.
Note: This may suggest overfitting, especially if the cross-validation score is not similarly high.

 ### Support Vector Regression (SVR):

CV_R2_Score: 0.847, the highest among all models.
R2: 0.786, indicating strong performance on the test data.
MAE & MSE: Low values, indicating good prediction accuracy.

 ### Ridge Regression:

CV_R2_Score: 0.318, better than simple linear regression.
R2: 0.612, indicating decent performance on the test data.
MAE & MSE: Improved over linear regression.

 ### Lasso and Elastic Net Regression:

CV_R2_Score: Negative values, indicating poor performance during cross-validation.
R2: Negative and close to zero, indicating poor performance on the test data.

 ### Random Forest Regression:

CV_R2_Score: 0.635, indicating good performance during cross-validation.
R2: 0.364, lower than expected performance on the test data.

 ### Gradient Boosting Regression:

CV_R2_Score: 0.648, indicating good performance during cross-validation.
R2: 0.287, indicating moderate performance on the test data.
 
 ### Principal Component Regression (PCR):

R2: 0.918, indicating very strong performance on the test data.
CV_R2_Score: 0.168, indicating potential overfitting.
 
 ### K-Nearest Neighbors Regression:

CV_R2_Score: Negative value, indicating poor performance during cross-validation.
R2: 0.322, indicating moderate performance on the test data.
 
 ### Neural Network Regression:

CV_R2_Score: 0.378, indicating decent performance during cross-validation.
R2: 0.405, indicating decent performance on the test data.
 
 ### Bayesian Regression:

CV_R2_Score: 0.333, indicating decent performance during cross-validation.
R2: 0.630, indicating good performance on the test data.
 
 ### Partial Least Squares Regression (PLS):

CV_R2_Score: 0.349, indicating decent performance during cross-validation.
R2: 0.656, indicating good performance on the test data.
 
## Conclusion:

 ### Best Performing Models:

Polynomial Regression and Support Vector Regression (SVR) show the best performance on the test data with very high R² scores.
PCR also shows a high R² score, but its cross-validation score suggests potential overfitting.

 ### Moderately Performing Models:

Ridge Regression, Bayesian Regression, and PLS show good performance with reasonable R² scores.
Poor Performing Models:

Lasso Regression, Elastic Net Regression, and K-Nearest Neighbors Regression show poor performance with negative or low R² scores.

### Caution with Overfitting:

Polynomial Regression shows perfect fit on the test data but might be overfitting, as indicated by the disparity between its cross-validation score and test score.
PCR shows high test performance but much lower cross-validation performance, suggesting overfitting.
Given the results, Support Vector Regression (SVR) appears to be the most reliable model with consistently high performance across both cross-validation and test data.

# Analysis of Cross Validation:


## Support Vector Regression (SVR):

CV_R2_Score: 0.846930 (high cross-validation score, indicating strong performance across folds)
R2: 0.785666 (high test score, indicating consistent performance on unseen data)

## Ridge Regression:

CV_R2_Score: 0.317766 (moderate cross-validation score)
R2: 0.612083 (good test score, indicating improved performance over linear regression)

## Polynomial Regression:

CV_R2_Score: 0.217806 (moderate cross-validation score, indicating potential overfitting)
R2: 1.000000 (perfect test score, likely overfitting)

## Principal Component Regression (PCR):

CV_R2_Score: 0.167647 (moderate cross-validation score)
R2: 0.918466 (high test score, indicating potential overfitting)

## Recommendations:

High Performing Models: SVR, Ridge Regression, and Bayesian Regression are strong candidates for further tuning and deployment due to their consistent performance.

Overfitting Concerns: Polynomial Regression and PCR show signs of overfitting despite their high test scores. These models should be used cautiously.

Further Tuning: Additional hyperparameter tuning and feature engineering could improve the performance of models like Random Forest and Gradient Boosting.

# Best Model Selection

## We will use GridSearchCV to perform hyperparameter tuning for each of these models.

Hyperparameter Tuning for Top 3 Models
1. Support Vector Regression (SVR)
2. Ridge Regression
3. Bayesian Regression

In [10]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.svm import SVR
from sklearn.linear_model import Ridge, BayesianRidge
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
import joblib

## Parameter List for the Model which will be tuned 

In [18]:
from scipy.stats import uniform
# Define hyperparameter grids for each model

param_grid_svr = {
    'C': [0.1, 1, 10, 100],
    'gamma': ['scale', 'auto'],
    'kernel': ['rbf', 'poly', 'sigmoid']
}

param_grid_ridge = {
    'alpha': [0.1, 1, 10, 100],
    'solver': ['auto', 'svd', 'cholesky', 'lsqr', 'sag']
}

param_grid_bayesian = {
    'n_iter': [100, 200, 300, 400, 500],
    'alpha_1': [1e-6, 1e-5, 1e-4],
    'alpha_2': [1e-6, 1e-5, 1e-4],
    'lambda_1': [1e-6, 1e-5, 1e-4],
    'lambda_2': [1e-6, 1e-5, 1e-4]
}


In [20]:
from sklearn.svm import SVR
from sklearn.linear_model import Ridge, BayesianRidge


In [21]:
# Initialize models
svr = SVR()
ridge = Ridge()  # Not `RidgeRegression`
bayesian = BayesianRidge()  # N

## Performing the Grid Search CV to find the best parameter for the models

In [22]:
# Perform grid search for each model
grid_svr = GridSearchCV(svr, param_grid_svr, cv=5, scoring='r2', n_jobs=-1, verbose=2)
grid_ridge = GridSearchCV(ridge, param_grid_ridge, cv=5, scoring='r2', n_jobs=-1, verbose=2)
grid_bayesian = GridSearchCV(bayesian, param_grid_bayesian, cv=5, scoring='r2', n_jobs=-1, verbose=2)

In [23]:
# Fit grid searches
grid_svr.fit(X_train, y_train)
grid_ridge.fit(X_train, y_train)
grid_bayesian.fit(X_train, y_train)

Fitting 5 folds for each of 24 candidates, totalling 120 fits
Fitting 5 folds for each of 20 candidates, totalling 100 fits
Fitting 5 folds for each of 405 candidates, totalling 2025 fits


ValueError: Invalid parameter 'n_iter' for estimator BayesianRidge(). Valid parameters are: ['alpha_1', 'alpha_2', 'alpha_init', 'compute_score', 'copy_X', 'fit_intercept', 'lambda_1', 'lambda_2', 'lambda_init', 'max_iter', 'tol', 'verbose'].

## Checking the best Parameters

In [15]:
# Print best parameters and scores for each model
print("Best parameters for SVR:", grid_svr.best_params_)
print("Best cross-validation R² score for SVR:", grid_svr.best_score_)

print("Best parameters for Ridge:", grid_ridge.best_params_)
print("Best cross-validation R² score for Ridge:", grid_ridge.best_score_)

print("Best parameters for Bayesian:", grid_bayesian.best_params_)
print("Best cross-validation R² score for Bayesian:", grid_bayesian.best_score_)

Best parameters for SVR: {'C': 10, 'gamma': 'auto', 'kernel': 'rbf'}
Best cross-validation R² score for SVR: 0.8411548315844988
Best parameters for Ridge: {'alpha': 10, 'solver': 'lsqr'}
Best cross-validation R² score for Ridge: 0.3074171887310845


AttributeError: 'GridSearchCV' object has no attribute 'best_params_'

## Now using the best parameters to tune the model 

In [36]:
# Evaluate the best models on the test set
best_svr = grid_svr.best_estimator_
best_ridge = grid_ridge.best_estimator_
best_bayesian = grid_bayesian.best_estimator_

models = [("SVR", best_svr), ("Ridge", best_ridge), ("Bayesian", best_bayesian)]

results = {}

for name, model in models:
    y_pred = model.predict(X_test)
    
    # Calculate metrics
    mae = mean_absolute_error(y_test, y_pred)
    mse = mean_squared_error(y_test, y_pred)
    r2 = r2_score(y_test, y_pred)
    
    # Store the results
    results[name] = {
        "MAE": mae,
        "MSE": mse,
        "R2": r2
    }

    # Save the model
    joblib.dump(model, f"{name}_best_model.pkl")

# Create a DataFrame to display the results
results_df = pd.DataFrame(results).T
print(results_df)

# Save the results to a CSV file
results_df.to_csv('hypertuned_model_results.csv')

               MAE       MSE        R2
SVR       0.269002  0.199971  0.778042
Ridge     0.466373  0.358771  0.601783
Bayesian  0.435209  0.333465  0.629872


# Additional feature Engineering

In [37]:
from sklearn.preprocessing import PolynomialFeatures

# Generate polynomial features
poly = PolynomialFeatures(degree=2, include_bias=False)
X_poly = poly.fit_transform(X_scaled)

# Split the data into training and testing sets
X_train_poly, X_test_poly, y_train_poly, y_test_poly = train_test_split(X_poly, y, test_size=0.2, random_state=42)

# Evaluate with SVR
svr_poly = SVR(C=10, gamma='scale', kernel='rbf')  # Using best params from previous tuning
svr_poly.fit(X_train_poly, y_train_poly)
y_pred_poly = svr_poly.predict(X_test_poly)

# Calculate metrics
mae_poly = mean_absolute_error(y_test_poly, y_pred_poly)
mse_poly = mean_squared_error(y_test_poly, y_pred_poly)
r2_poly = r2_score(y_test_poly, y_pred_poly)

print(f"SVR with Polynomial Features - MAE: {mae_poly}, MSE: {mse_poly}, R2: {r2_poly}")


SVR with Polynomial Features - MAE: 0.28470374674347354, MSE: 0.2071481868718464, R2: 0.7700766161838215


# Advanced Ensembled Method

In [38]:
import xgboost as xgb

# Define and train XGBoost model
xgb_model = xgb.XGBRegressor(n_estimators=100, learning_rate=0.1, max_depth=5, random_state=42)
xgb_model.fit(X_train, y_train)
y_pred_xgb = xgb_model.predict(X_test)

# Calculate metrics
mae_xgb = mean_absolute_error(y_test, y_pred_xgb)
mse_xgb = mean_squared_error(y_test, y_pred_xgb)
r2_xgb = r2_score(y_test, y_pred_xgb)

print(f"XGBoost Model - MAE: {mae_xgb}, MSE: {mse_xgb}, R2: {r2_xgb}")


XGBoost Model - MAE: 0.3805426571051523, MSE: 0.5474853138676409, R2: 0.39232064805864386


# Stacking Models (Ensemble )

In [51]:
import numpy as np
from sklearn.ensemble import StackingRegressor
from sklearn.linear_model import LinearRegression, Ridge
from sklearn.svm import SVR
from sklearn.linear_model import BayesianRidge
from sklearn.model_selection import cross_val_score, KFold
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

# Define base models
base_models = [
    ('ridge', Ridge(alpha=10)),
    ('svr', SVR(C=10, gamma='scale', kernel='rbf')),
    ('bayesian', BayesianRidge())
]

# Define stacking model
stacking_model = StackingRegressor(
    estimators=base_models,
    final_estimator=LinearRegression()
)

# Define KFold cross-validator
kf = KFold(n_splits=5, shuffle=True, random_state=42)

# Evaluate model using cross-validation
cv_mae_scores = cross_val_score(stacking_model, X_train, y_train, cv=kf, scoring='neg_mean_absolute_error')
cv_mse_scores = cross_val_score(stacking_model, X_train, y_train, cv=kf, scoring='neg_mean_squared_error')
cv_r2_scores = cross_val_score(stacking_model, X_train, y_train, cv=kf, scoring='r2')

# Print cross-validation results for each fold
print("Cross-validation MAE scores for each fold:", -cv_mae_scores)
print("Cross-validation MSE scores for each fold:", -cv_mse_scores)
print("Cross-validation R2 scores for each fold:", cv_r2_scores)

# Print mean cross-validation scores
print(f"Mean cross-validated MAE: {-cv_mae_scores.mean()}")
print(f"Mean cross-validated MSE: {-cv_mse_scores.mean()}")
print(f"Mean cross-validated R2: {cv_r2_scores.mean()}")

# Train stacking model on full training set
stacking_model.fit(X_train, y_train)
y_pred_stack = stacking_model.predict(X_test)

# Calculate metrics on test set
mae_stack = mean_absolute_error(y_test, y_pred_stack)
mse_stack = mean_squared_error(y_test, y_pred_stack)
r2_stack = r2_score(y_test, y_pred_stack)

print(f"Test Set - Stacking Model - MAE: {mae_stack}, MSE: {mse_stack}, R2: {r2_stack}")


Cross-validation MAE scores for each fold: [0.22542481 0.16927933 0.28900586 0.30879518 0.14519045]
Cross-validation MSE scores for each fold: [0.10238496 0.06096286 0.13254362 0.12942435 0.02374588]
Cross-validation R2 scores for each fold: [0.87587693 0.67114169 0.90873866 0.76880754 0.96013668]
Mean cross-validated MAE: 0.22753912841317528
Mean cross-validated MSE: 0.08981233463984364
Mean cross-validated R2: 0.8369403000290113
Test Set - Stacking Model - MAE: 0.21034456500762985, MSE: 0.09506965002700721, R2: 0.8944777844183961


## More Experiment with Ensemble Models

In [24]:
from sklearn.linear_model import LinearRegression, Ridge, Lasso, ElasticNet, BayesianRidge
from sklearn.svm import SVR
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.neighbors import KNeighborsRegressor
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
from sklearn.ensemble import StackingRegressor
import joblib

# Load the final feature engineered data
df_final = pd.read_csv('final_feature_engineered_data.csv')

# Separate the features and target
X = df_final.drop(columns=['Log_Arsenic'])
y = df_final['Log_Arsenic']

# Standardize the features
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.2, random_state=42)

# Define a list of base models
base_models = [
    ('ridge', Ridge(alpha=10)),
    ('svr', SVR(C=10, gamma='scale', kernel='rbf')),
    ('bayesian', BayesianRidge()),
    ('rf', RandomForestRegressor(random_state=42)),
    ('gb', GradientBoostingRegressor(random_state=42)),
    ('knn', KNeighborsRegressor())
]

# Define a list of meta-models
meta_models = [
    ('linear', LinearRegression()),
    ('ridge', Ridge(alpha=10)),
    ('lasso', Lasso(alpha=0.1)),
    ('elasticnet', ElasticNet(alpha=0.1))
]


In [26]:
import itertools
from sklearn.ensemble import StackingRegressor, RandomForestRegressor, GradientBoostingRegressor
from sklearn.linear_model import Ridge, BayesianRidge, LinearRegression, Lasso, ElasticNet
from sklearn.svm import SVR
from sklearn.neighbors import KNeighborsRegressor
from sklearn.model_selection import KFold, cross_val_score
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
import joblib
import pandas as pd

# Define a list of base models
base_models = [
    ('ridge', Ridge(alpha=10)),
    ('svr', SVR(C=10, gamma='scale', kernel='rbf')),
    ('bayesian', BayesianRidge()),
    ('rf', RandomForestRegressor(random_state=42)),
    ('gb', GradientBoostingRegressor(random_state=42)),
    ('knn', KNeighborsRegressor())
]

# Define a list of meta-models
meta_models = [
    ('linear', LinearRegression()),
    ('ridge', Ridge(alpha=10)),
    ('lasso', Lasso(alpha=0.1)),
    ('elasticnet', ElasticNet(alpha=0.1))
]

# Dictionary to store the results
results = {}

# Define k-fold cross-validation
kf = KFold(n_splits=5, shuffle=True, random_state=42)

# Iterate through each combination of base models and meta-models
for meta_name, meta_model in meta_models:
    for i in range(1, len(base_models) + 1):
        for base_combination in itertools.combinations(base_models, i):
            # Ensure base_combination is a list
            base_combination_list = list(base_combination)
            
            # Define the stacking model
            stacking_model = StackingRegressor(
                estimators=base_combination_list,
                final_estimator=meta_model,
                cv=kf,
                n_jobs=-1
            )
            
            # Perform k-fold cross-validation
            cv_results = cross_val_score(stacking_model, X_train, y_train, cv=kf, scoring='r2', n_jobs=-1)
            mean_cv_r2 = cv_results.mean()
            std_cv_r2 = cv_results.std()
            
            # Train the stacking model
            stacking_model.fit(X_train, y_train)
            y_pred_stack = stacking_model.predict(X_test)
            
            # Calculate metrics
            mae_stack = mean_absolute_error(y_test, y_pred_stack)
            mse_stack = mean_squared_error(y_test, y_pred_stack)
            r2_stack = r2_score(y_test, y_pred_stack)
            
            # Store the results
            model_name = f"STACKING_{meta_name.upper()}_WITH_{'_'.join([name.upper() for name, _ in base_combination])}"
            results[model_name] = {
                "CV_Mean_R2": mean_cv_r2,
                "CV_Std_R2": std_cv_r2,
                "MAE": mae_stack,
                "MSE": mse_stack,
                "R2": r2_stack
            }

            # Save the model
            joblib.dump(stacking_model, f"{model_name}.pkl")

# Create a DataFrame to display the results
results_df = pd.DataFrame(results).T
print(results_df)

# Save the results to a CSV file
results_df.to_csv('ensemble_model_comparison_results.csv')


                                                    CV_Mean_R2  CV_Std_R2  \
STACKING_LINEAR_WITH_RIDGE                            0.081148   0.392351   
STACKING_LINEAR_WITH_SVR                              0.820242   0.140375   
STACKING_LINEAR_WITH_BAYESIAN                         0.013654   0.257066   
STACKING_LINEAR_WITH_RF                               0.741289   0.136416   
STACKING_LINEAR_WITH_GB                               0.740399   0.222841   
...                                                        ...        ...   
STACKING_ELASTICNET_WITH_RIDGE_SVR_BAYESIAN_GB_KNN    0.818225   0.097971   
STACKING_ELASTICNET_WITH_RIDGE_SVR_RF_GB_KNN          0.838978   0.088808   
STACKING_ELASTICNET_WITH_RIDGE_BAYESIAN_RF_GB_KNN     0.771861   0.140312   
STACKING_ELASTICNET_WITH_SVR_BAYESIAN_RF_GB_KNN       0.818597   0.097804   
STACKING_ELASTICNET_WITH_RIDGE_SVR_BAYESIAN_RF_...    0.818279   0.097955   

                                                         MAE       MSE  \
S

In [27]:
results_df

CV_Mean_R2  CV_Std_R2  \
STACKING_LINEAR_WITH_RIDGE                            0.081148   0.392351   
STACKING_LINEAR_WITH_SVR                              0.820242   0.140375   
STACKING_LINEAR_WITH_BAYESIAN                         0.013654   0.257066   
STACKING_LINEAR_WITH_RF                               0.741289   0.136416   
STACKING_LINEAR_WITH_GB                               0.740399   0.222841   
...                                                        ...        ...   
STACKING_ELASTICNET_WITH_RIDGE_SVR_BAYESIAN_GB_KNN    0.818225   0.097971   
STACKING_ELASTICNET_WITH_RIDGE_SVR_RF_GB_KNN          0.838978   0.088808   
STACKING_ELASTICNET_WITH_RIDGE_BAYESIAN_RF_GB_KNN     0.771861   0.140312   
STACKING_ELASTICNET_WITH_SVR_BAYESIAN_RF_GB_KNN       0.818597   0.097804   
STACKING_ELASTICNET_WITH_RIDGE_SVR_BAYESIAN_RF_...    0.818279   0.097955   

                                                         MAE       MSE  \
STACKING_LINEAR_WITH_RIDGE                          0.491528  0.381131   
STACKING_LINEAR_WITH_SVR                            0.202634  0.105406   
STACKING_LINEAR_WITH_BAYESIAN                       0.618847  0.556424   
STACKING_LINEAR_WITH_RF                             0.381499  0.481085   
STACKING_LINEAR_WITH_GB                             0.397857  0.615138   
...                                                      ...       ...   
STACKING_ELASTICNET_WITH_RIDGE_SVR_BAYESIAN_GB_KNN  0.316764  0.303450   
STACKING_ELASTICNET_WITH_RIDGE_SVR_RF_GB_KNN        0.319557  0.317061   
STACKING_ELASTICNET_WITH_RIDGE_BAYESIAN_RF_GB_KNN   0.420550  0.555156   
STACKING_ELASTICNET_WITH_SVR_BAYESIAN_RF_GB_KNN     0.319550  0.317048   
STACKING_ELASTICNET_WITH_RIDGE_SVR_BAYESIAN_RF_...  0.319552  0.317055   

                                                          R2  
STACKING_LINEAR_WITH_RIDGE                          0.576965  
STACKING_LINEAR_WITH_SVR                            0.883005  
STACKING_LINEAR_WITH_BAYESIAN                       0.382399  
STACKING_LINEAR_WITH_RF                             0.466021  
STACKING_LINEAR_WITH_GB                             0.317230  
...                                                      ...  
STACKING_ELASTICNET_WITH_RIDGE_SVR_BAYESIAN_GB_KNN  0.663186  
STACKING_ELASTICNET_WITH_RIDGE_SVR_RF_GB_KNN        0.648079  
STACKING_ELASTICNET_WITH_RIDGE_BAYESIAN_RF_GB_KNN   0.383807  
STACKING_ELASTICNET_WITH_SVR_BAYESIAN_RF_GB_KNN     0.648094  
STACKING_ELASTICNET_WITH_RIDGE_SVR_BAYESIAN_RF_...  0.648086  

[252 rows x 5 columns]

# Select the best ensemble model:

In [28]:
import pandas as pd

# Load the results DataFrame
results_df = pd.read_csv('ensemble_model_comparison_results.csv')

# Select the best model based on test R² score
best_models_r2 = results_df.sort_values(by='R2', ascending=False)
best_model_r2 = best_models_r2.iloc[0]
print("Best Model based on test R² score:")
print(best_model_r2)

# Select the best model based on Cross-Validation Mean R² score
best_models_cv_r2 = results_df.sort_values(by='CV_Mean_R2', ascending=False)
best_model_cv_r2 = best_models_cv_r2.iloc[0]
print("\nBest Model based on Cross-Validation Mean R² score:")
print(best_model_cv_r2)

# Optionally, display the top 5 models based on each criterion
top_5_models_r2 = best_models_r2.head(5)
print("\nTop 5 Models based on test R² score:")
print(top_5_models_r2)

top_5_models_cv_r2 = best_models_cv_r2.head(5)
print("\nTop 5 Models based on Cross-Validation Mean R² score:")
print(top_5_models_cv_r2)


Best Model based on test R² score:
Unnamed: 0    STACKING_LINEAR_WITH_SVR_RF
CV_Mean_R2                       0.794076
CV_Std_R2                        0.183497
MAE                              0.182704
MSE                              0.073791
R2                               0.918096
Name: 12, dtype: object

Best Model based on Cross-Validation Mean R² score:
Unnamed: 0    STACKING_LINEAR_WITH_SVR_KNN
CV_Mean_R2                        0.879771
CV_Std_R2                          0.10028
MAE                               0.265883
MSE                               0.124153
R2                                0.862197
Name: 14, dtype: object

Top 5 Models based on test R² score:
                                       Unnamed: 0  CV_Mean_R2  CV_Std_R2  \
12                    STACKING_LINEAR_WITH_SVR_RF    0.794076   0.183497   
41     STACKING_LINEAR_WITH_RIDGE_SVR_BAYESIAN_RF    0.639876   0.283577   
56  STACKING_LINEAR_WITH_RIDGE_SVR_BAYESIAN_RF_GB    0.404885   0.680023   
31          

In [29]:
top_5_models_cv_r2


Unnamed: 0  CV_Mean_R2  CV_Std_R2       MAE  \
14             STACKING_LINEAR_WITH_SVR_KNN    0.879771   0.100280  0.265883   
202         STACKING_ELASTICNET_WITH_SVR_GB    0.845409   0.089937  0.316762   
225     STACKING_ELASTICNET_WITH_SVR_GB_KNN    0.845409   0.089937  0.316762   
243  STACKING_ELASTICNET_WITH_SVR_RF_GB_KNN    0.845243   0.089816  0.319550   
223      STACKING_ELASTICNET_WITH_SVR_RF_GB    0.845243   0.089816  0.319550   

          MSE        R2  
14   0.124153  0.862197  
202  0.303447  0.663190  
225  0.303447  0.663190  
243  0.317048  0.648094  
223  0.317048  0.648094

In [30]:
top_5_models_r2

Unnamed: 0  CV_Mean_R2  CV_Std_R2  \
12                    STACKING_LINEAR_WITH_SVR_RF    0.794076   0.183497   
41     STACKING_LINEAR_WITH_RIDGE_SVR_BAYESIAN_RF    0.639876   0.283577   
56  STACKING_LINEAR_WITH_RIDGE_SVR_BAYESIAN_RF_GB    0.404885   0.680023   
31           STACKING_LINEAR_WITH_SVR_BAYESIAN_RF    0.728839   0.282440   
34                 STACKING_LINEAR_WITH_SVR_RF_GB    0.745078   0.201218   

         MAE       MSE        R2  
12  0.182704  0.073791  0.918096  
41  0.198073  0.078027  0.913394  
56  0.192374  0.087071  0.903356  
31  0.213909  0.089995  0.900110  
34  0.181319  0.091967  0.897922

# Choosing the best model 

## Performance of the New Stacking Model:
Cross-validation MAE: 0.2275
Cross-validation MSE: 0.0898
Cross-validation R2: 0.8369
Test MAE: 0.2103
Test MSE: 0.0951
Test R2: 0.8945


## Previous Best Models from the Tables:

### Top Models Based on Cross-Validation Mean R² Score:

Stacking_linear_with_svr_knn
CV_Mean_R2: 0.8798
CV_Std_R2: 0.1003
MAE: 0.2659
MSE: 0.1242
R2: 0.8622

### Top Models Based on Test R² Score:

Stacking_linear_with_svr_rf
CV_Mean_R2: 0.7941
CV_Std_R2: 0.1835
MAE: 0.1827
MSE: 0.0738
R2: 0.9181

## Comparison:
### Cross-Validation Metrics:

New Stacking Model: CV_R2: 0.8369, CV_MAE: 0.2275, CV_MSE: 0.0898
Stacking_linear_with_svr_knn: CV_R2: 0.8798, CV_MAE: 0.2659, CV_MSE: 0.1242
Stacking_linear_with_svr_rf: CV_R2: 0.7941, CV_MAE: 0.1827, CV_MSE: 0.0738


### Test Set Metrics:

New Stacking Model: R2: 0.8945, MAE: 0.2103, MSE: 0.0951
Stacking_linear_with_svr_knn: R2: 0.8622, MAE: 0.2659, MSE: 0.1242
Stacking_linear_with_svr_rf: R2: 0.9181, MAE: 0.1827, MSE: 0.0738

## Conclusion:

## Best Model Based on Test Set R² Score:

### Stacking_linear_with_svr_rf: Test R²: 0.9181, Test MAE: 0.1827, Test MSE: 0.0738

### Best Model Based on Cross-Validation Mean R² Score:

### Stacking_linear_with_svr_knn: CV_R2: 0.8798, CV_MAE: 0.2659, CV_MSE: 0.1242

New Stacking Model:


Test R²: 0.8945 (second highest)
CV_R2: 0.8369 (second highest among provided models)

The new stacking model performs very well in terms of test set performance and has solid cross-validation results.

### Given the performance metrics, Stacking_linear_with_svr_rf remains the best model overall due to its highest R² score on the test set. 

However, the new stacking model with base models ridge, svr, and bayesian, and a LinearRegression final estimator also performs exceptionally well and could be a very strong alternative.

## Final Decision:
Primary Choice: Stacking_linear_with_svr_rf

Strong Alternative: The new stacking model (ridge, svr, bayesian with LinearRegression as final estimator)

# Saving the best Model

In [31]:
import joblib
import pandas as pd

# Assuming 'Stacking_linear_with_svr_rf' is the best model based on the analysis
best_model_name = 'Stacking_linear_with_svr_rf'

# Load the best model
best_model = joblib.load(f"{best_model_name}.pkl")

# Save the best model with a new name for clarity
joblib.dump(best_model, 'best_stacking_model.pkl')

# Document the final results
best_model_results = results_df.loc[results_df.index == best_model_name]
print("Best Model Results:")
print(best_model_results)

# Save the best model results to a CSV file
best_model_results.to_csv('best_model_results.csv', index=False)

# Optionally, save the strong alternative model
alternative_model_name = 'Stacking_new_model_ridge_svr_bayesian'
joblib.dump(stacking_model, f"{alternative_model_name}.pkl")


Best Model Results:
Empty DataFrame
Columns: [Unnamed: 0, CV_Mean_R2, CV_Std_R2, MAE, MSE, R2]
Index: []


['Stacking_new_model_ridge_svr_bayesian.pkl']